autodiff/forward/real/real.hpp

```diff
@@ -0,0 +1,835 @@
+//
+// BINARY OPERATOR -(Real, Real), -(Real, Number), -(Number, Real)
+//
+//=====================================================================================================================
+template<size_t N, typename T>
+auto operator-(Real<N, T> x, const Real<N, T>& y)
+{
+    return x -= y;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+auto operator-(Real<N, T> x, const U& y)
+{
+    return x -= y;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+auto operator-(const U& x, Real<N, T> y)
+{
+    For<0, N + 1>([&](auto i) constexpr { y[i] = static_cast<T>(x) - y[i]; });
+    return y;
+}
+
+//=====================================================================================================================
+//
+// BINARY OPERATOR *(Real, Real), *(Real, Number), *(Number, Real)
+//
+//=====================================================================================================================
+
+template<size_t N, typename T>
+auto operator*(Real<N, T> x, const Real<N, T>& y)
+{
+    return x *= y;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+auto operator*(Real<N, T> x, const U& y)
+{
+    return x *= y;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+auto operator*(const U& x, Real<N, T> y)
+{
+    return y *= x;
+}
+
+//=====================================================================================================================
+//
+// BINARY OPERATOR /(Real, Real), /(Real, Number), /(Number, Real)
+//
+//=====================================================================================================================
+
+template<size_t N, typename T>
+auto operator/(Real<N, T> x, const Real<N, T>& y)
+{
+    return x /= y;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+auto operator/(Real<N, T> x, const U& y)
+{
+    return x /= y;
+}
+
+template<size_t N, typename T, typename U, EnableIf<isNumber<U>>...>
+auto operator/(const U& x, Real<N, T> y)
+{
+    Real<N, T> z = x;
+    return z /= y;
+}
+
+//=====================================================================================================================
+//
+// EXPONENTIAL AND LOGARITHMIC FUNCTIONS
+//
+//=====================================================================================================================
+
+template<size_t N, typename T>
+constexpr auto exp(const Real<N, T>& x)
+{
+    Real<N, T> expx;
+    expx[0] = exp(x[0]);
+    For<1, N + 1>([&](auto i) constexpr {
+        expx[i] = Sum<0, i>([&](auto j) constexpr {
+            constexpr auto c = BinomialCoefficient<i.index - 1, j.index>;
+            return c * x[i - j] * expx[j];
+        });
+    });
+    return expx;
+}
+
+template<size_t N, typename T>
+constexpr auto log(const Real<N, T>& x)
+{
+    assert(x[0] != 0 && "autodiff: log(x) has undefined value and derivatives when x = 0");
```

Comment:
It can be not only a strict zero. What do you think about checking small values?

Reply:
I propose we stick with strict zero; otherwise we need to define what is a small value, which is not always intuitive or possible.